script/import_scripts/smf2.rb

```diff
@@ -266,11 +269,121 @@ def guardian
           )
         end
       end
-      post[:raw] = convert_message_body(message[:body], attachments, ignore_quotes: ignore_quotes)
+      begin
+        post[:raw] = convert_message_body(message[:body], attachments, ignore_quotes: ignore_quotes)
+      rescue StandardError
+        post[:raw] = "-- MESSAGE SKIPPED --"
+      end
       next post
     end
   end
 
+  def import_personal_posts
+    puts "Loading pm mapping..."
+
+    @pm_mapping = {}
+
+    Topic
+      .joins(:topic_allowed_users)
+      .where(archetype: Archetype.private_message)
+      .where("title NOT ILIKE 'Re:%'")
+      .group(:id)
+      .order(:id)
+      .pluck(
+        "string_agg(topic_allowed_users.user_id::text, ',' ORDER BY topic_allowed_users.user_id), title, topics.id",
+        )
+      .each do |users, title, topic_id|
+      @pm_mapping[users] ||= {}
+      @pm_mapping[users][title] ||= []
+      @pm_mapping[users][title] << topic_id
+    end
+
+    puts "", "Importing personal posts..."
+
+    last_post_id = -1
+    total =
+      query(
+        "SELECT COUNT(*) count FROM smf_personal_messages WHERE deleted_by_sender = 0",
+        as: :single)
+
+    batches(BATCH_SIZE) do |offset|
+      posts = query(<<~SQL, as: :array)
+        SELECT id_pm
+             , id_member_from
+             , msgtime
+             , subject
+             , body
+             , (SELECT GROUP_CONCAT(id_member) FROM smf_pm_recipients r WHERE r.id_pm = pm.id_pm) recipients
+          FROM smf_personal_messages pm
+         WHERE deleted_by_sender = 0
+           AND id_pm > #{last_post_id}
+         ORDER BY id_pm
+         LIMIT #{BATCH_SIZE}
+      SQL
+
+      break if posts.empty?
+
+      last_post_id = posts[-1][:id_pm]
+      post_ids = posts.map { |p| "pm-#{p[:id_pm]}" }
+
+      next if all_records_exist?(:post, post_ids)
+
+      create_posts(posts, total: total, offset: offset) do |p|
+        next unless user_id = user_id_from_imported_user_id(p[:id_member_from])
+        next if p[:recipients].blank?
+        recipients =
+          p[:recipients].split(",").map { |id| user_id_from_imported_user_id(id) }.compact.uniq
+        next if recipients.empty?
+
+        id = "pm-#{p[:id_pm]}"
+        puts id
+        next if post_id_from_imported_post_id(id)
+
+        post = {
+          id: id,
+          created_at: Time.at(p[:msgtime]),
+          user_id: user_id,
+        }
+        begin
+          post[:raw] = convert_message_body(p[:body])
+        rescue StandardError
+          post[:raw] = "-- MESSAGE SKIPPED --"
+        end
+
+
+        users = (recipients + [user_id]).sort.uniq.join(",")
+        title = decode_entities(p[:subject])
+
+        if topic_id = find_pm_topic_id(users, title)
```

Comment:
Is this really the only way of finding the topic for a message?

I'm not familiar with the DB schema of SMF2, but ChatGPT tells me that the `smf_personal_messages` is supposed to have a `id_pm_head` column. Would you mind verifying if that's correct? If so, it might make it easier to group messages.

> In SMF2, the `smf_personal_messages` table contains information about private messages (PMs), but it doesn't directly store conversation threads in a structured way. However, you can find out which messages belong to the same conversation by using the `id_pm_head` column. Here's how it works:
> 
> 1. **`id_pm_head`**: This column acts as the identifier for a conversation thread. All messages that share the same `id_pm_head` are part of the same conversation. When a new conversation starts, the message’s `id_pm_head` will be set to its `id_pm`. Replies to this message will share the same `id_pm_head`, linking them to the original message as part of the conversation.
> 
> 2. **`id_pm`**: Each message has a unique identifier, `id_pm`. For the first message in a conversation, `id_pm` will be the same as `id_pm_head`. Replies will have different `id_pm` values but share the same `id_pm_head` as the initial message in the thread.

Reply:
I used the `id_pm_head` value in a different importer script I wrote (in Python/Django, since that's what I'm more familiar with) and it works but this is what was used in `import_scripts/smf1.rb` and I left it since it works just as well.

Reply:
Okay, not optimal, but let's use it for now, unless you want to rework this. If not, please add a comment to `find_pm_topic_id` that this could be improved by using `id_pm_head`.